File: http-endpoint-service/src/main/scala/http_endpoint_service/api/HttpEndpointServiceApi.scala
package http_endpoint_service.api

import java.util.concurrent.Future

import akka.actor.ActorSystem
import akka.http.scaladsl.model.StatusCodes
import akka.http.scaladsl.server.Directives._
import akka.http.scaladsl.server.Route
import akka.stream.Materializer
import de.heikoseeberger.akkahttpplayjson.PlayJsonSupport
import http_endpoint_service.domain.GPSUpdate
import ch.megard.akka.http.cors.scaladsl.CorsDirectives._
import org.apache.kafka.clients.producer.{KafkaProducer, ProducerRecord, RecordMetadata}
import play.api.libs.json.Json

import scala.concurrent.ExecutionContext

/**
  * Create endpoints and send all incoming messages to the actor given.
  */
class HttpEndpointServiceApi(producer: KafkaProducer[String, String], kafkaTopicOut: String)(
    implicit
    system: ActorSystem,
    mat: Materializer,
    ex: ExecutionContext
) extends PlayJsonSupport {

  import GPSUpdate._

  val route: Route = cors() {
    pathPrefix("health") {
      get {
        complete(StatusCodes.OK)
      }
    } ~
      pathPrefix("addGPSUpdate") {
        post {
          entity(as[GPSUpdate]) { gpsUpdate =>
            println(s"received: $gpsUpdate")

            // Send the gps update to the actor who puts it on Kafka
            produceItem(gpsUpdate)
            complete(StatusCodes.OK)
          }
        }
      } ~
      pathPrefix("addMultipleUpdates") {
        post {
          entity(as[Seq[GPSUpdate]]) { gpsUpdates =>
            println(s"received $gpsUpdates")
            // Send each update as a separate message to the actor putting it on Kafka
            gpsUpdates.foreach(gpsUpdate => produceItem(gpsUpdate))
            complete(StatusCodes.OK)
          }
        }
      }
  }

  def produceItem(loc: GPSUpdate): Future[RecordMetadata] = {
    println(s"Producing: $loc")
    producer.send(new ProducerRecord[String, String](kafkaTopicOut, Json.toJson(loc).toString))
  }
}

File: java/http-endpoint-service/src/main/java/api/HttpEndpointServiceApi.java
package api;

import akka.http.javadsl.model.StatusCodes;
import akka.http.javadsl.server.AllDirectives;
import akka.http.javadsl.server.Route;
import domain.GPSUpdate;
import org.apache.kafka.clients.producer.KafkaProducer;
import org.apache.kafka.clients.producer.ProducerRecord;
import akka.http.javadsl.marshallers.jackson.Jackson;

import java.util.LinkedHashMap;
import java.util.List;

import static ch.megard.akka.http.cors.javadsl.CorsDirectives.cors;

public class HttpEndpointServiceApi extends AllDirectives {

    private KafkaProducer<String, String> producer;
    private String kafkaTopicOut;

    public HttpEndpointServiceApi(KafkaProducer<String, String> producer, String kafkaTopicOut) {
        this.kafkaTopicOut = kafkaTopicOut;
        this.producer = producer;
    }

    public Route getRoute() {

        return cors(() -> route(
            pathPrefix("health", () ->
                get(() -> complete(StatusCodes.OK))
            ),
            pathPrefix("addGPSUpdate", () ->
                post(() ->
                    entity(Jackson.unmarshaller(GPSUpdate.class), gpsUpdate -> {
                        System.out.println("received: " + gpsUpdate);

                        // Send the gps update to the actor who puts it on Kafka
                        produceItem(gpsUpdate);
                        return complete(StatusCodes.OK);
                    })
                )
            ),
            pathPrefix("addMultipleUpdates", () ->
                    post(() ->
                            entity(Jackson.unmarshaller(List.class), gpsUpdates -> {
                                // Send the gps update to the actor who puts it on Kafka
                                // Obviously you don't want the unmarshalling in this state on production, but it serves the demo purpose
                                // Unmarshalling a json array with the scaladsl is a bit easier
                                gpsUpdates.forEach(update -> {
                                    LinkedHashMap<String, Object> castUpdate = (LinkedHashMap<String, Object>) update;
                                    produceItem(
                                        new GPSUpdate(
                                            (String) castUpdate.get("uniqueId"),
                                            (double) castUpdate.get("latitude"),
                                            (double) castUpdate.get("longitude"),
                                            (long) ((int) castUpdate.get("timestamp"))
                                        )
                                    );
                                });
                                return complete(StatusCodes.OK);
                            })
                    )
            )
        ));
    }

    private void produceItem(GPSUpdate loc) {
        producer.send(new ProducerRecord<>(kafkaTopicOut, loc.toJson()));
    }
}
